exo/inference/pytorch/model/hf.py

```diff
@@ -0,0 +1,107 @@
+import torch
+import torch.nn as nn
+import numpy as np
+
+from transformers import AutoModelForCausalLM, LlamaConfig, DynamicCache, Cache
+from exo.inference.shard import Shard
+from exo.helpers import DEBUG
+from typing import Tuple
+
+from .utils import sample_logits
+
+class ShardedHuggingFaceModel(torch.nn.Module):
+    def __init__(self, shard: Shard):
+        super(ShardedHuggingFaceModel, self).__init__()
+
+        if DEBUG >= 2:
+            print(f"\nShardedHuggingFaceModel init with shard {shard}")
+
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        self.shard = shard
+
+        
+
+        # Load the model
+        self.full_model = AutoModelForCausalLM.from_pretrained(
+            shard.model_id,
+            torch_dtype=torch.float16 if torch.cuda.is_available() else torch.float32,
+            device_map="auto"
+        )
+
+        # using llamaconfig not working setting layers manually
+        layers = []
+        for i in range(shard.start_layer, shard.end_layer + 1):
+            layer = self.full_model.model.layers[i]
+
+            if DEBUG >= 2:
+                print(f"Loading layers[{i}]")
+
+            layers.append(layer)
+        
+        self.full_model.model.layers = nn.ModuleList(layers)
```

Comment:
What does the peak memory usage look like here? I'm not sure of the specifics of python if this is going to hold each layer twice. Not sure but perhaps setting them in place would be more memory efficient.

Reply:
They shouldn't be held twice as when the ensure_shard function is called in the infer_prompt or infer_tensor the  init class function is called which loads the needed layers each time depending on the shard. Will make sure about memory limits though and usage.